### ISSUE TITLE
Adding custom serializer to list generates autoload deprecation warning

### ISSUE THREAD

**Issue by notapatch** [rails/rails#37690]
Trying to use a new feature for Rails 6 [Introduce custom serializers to ActiveJob arguments](https://github.com/rails/rails/pull/30941) and I am getting a warning when I add a serializer to the `custom_serializer` list.

### Steps to reproduce

Included the 5 steps for setup before the actual issue under test in "testing"

### Setup

1. `rails new custom_serializer_warning`
2. `cd custom_serializer_warning && git add . && git commit -m "initial commit"`
3. `mkdir app/serializers && touch app/serializers/money_serializer.rb`
4. [copy money_serializer from active_job_basics docs](https://edgeguides.rubyonrails.org/active_job_basics.html#serializers). 
5. Add the serializer - Either of initializer file or directly into application.rb. The guide doesn't make it clear if there's a preferred way (both are in the example repo): 

a. Initializer file
```ruby
#config/initializers/custom_serializer.rb

Rails.application.config.active_job.custom_serializers << MoneySerializer
```

b. Adding to` application.rb` directly 
```ruby
 # application.rb

 class Application < Rails::Application
     # Initialize configuration defaults for originally generated Rails version.
     config.load_defaults 6.0
+    config.active_job.custom_serializers << MoneySerializer
   end
```
This way gives a different behaviour to (a) in that the loading error exception stops me going into rails console at all (see below)

[I have a minimal rails example](https://github.com/notapatch/custom_serializer_warning)

---
### Testing

1. `DISABLE_SPRING=1 rails test`

- Without the DISABLE_SPRING you only see it the first time. 

Included the `MoneySerializer` part of `Rails.autoloaders.log!`
```
DEPRECATION WARNING: Initialization autoloaded the constant MoneySerializer.
(see below for full message)
...
Please, check the "Autoloading and Reloading Constants" guide for solutions.
 (called from <main> at 
...
Zeitwerk@rails.main: autoload set for MoneySerializer, to be loaded from
 /.../custom_serializer_warning/app/serializers/money_serializer.rb
```

2. Rails console

When adding to custom serializer with the initializer file (a) it loads without warning (the autoload does not try to load it). When you do force it to use MoneySerializer it loads without warning.

```ruby
m = MoneySerializer
Zeitwerk@rails.main: constant MoneySerializer loaded from file 
/../custom_serializer_warning/app/serializers/money_serializer.rb
=> MoneySerializer
```

When adding to custom serializer with application.rb (b) it errors out

```ruby
uninitialized constant
      CustomSerializerWarning::Application::MoneySerializer (NameError)
```

### Variations which make no difference
This makes no difference to the warning
- Rails 6.0.1 / Rails master
- directory "serializers" doesn't matter I have created other directories below `app` without changing anything
- Full Serializer class  as with `MoneySerializer`  or an empty class `class MySerializer; end`  makes no difference.
- Adding serializer in `config/environments/test.rb ` it behaved like (a) adding via initializer file.
- `Rails.autoloaders.log! ` present / absent

### Expected behavior
When a custom serializer is added to custom serializers no warning is given. Or failing that I would expect guide to reassure that what I am doing shouldn't break in the near future.

### Actual behavior
```
DEPRECATION WARNING: Initialization autoloaded the constant <FooSerializer>.

Being able to do this is deprecated. Autoloading during initialization is going
to be an error condition in future versions of Rails.

Reloading does not reboot the application, and therefore code executed during
initialization does not run again. So, if you reload <FooSerializer>, for example,
the expected changes won't be reflected in that stale Class object.

This autoloaded constant has been unloaded.
```

### System configuration
Rails 6.0.1 (but also tested on master)
Ruby 2.6.5p114

### References
[Introduce custom serializers to ActiveJob arguments](https://github.com/rails/rails/pull/30941)

**Comment by fxn** [comment]
Hey, super quick response.

Problem is, if you edit that serializer and reload, the changes won't take effect because the class object is already stored in the custom serializers. It is a use case of cached class objects while the app boots, which does not play well with reloading. That is why we print a warning now, and why that won't be possible in the future (mod, that future has not arrived :).

You don't see the autoload in the logs while the app boots because Zeitwerk is configured and setup after the initializers run. Before that, classic is autoloading (a trade-off of the transition that we accepted because autoloading shouldn't happen during boot anyway). That is why Rails unloads the autoloaded constants in addition to print the warning.

One possible workaround by now would be to tell Zeitwerk to ignore that directory, and `require` de file:

```ruby
# config/initializers/custom_serializer.rb

Rails.autoloaders.main.ignore("#{Rails.root}/app/serializers")

require "money_serializer"
Rails.application.config.active_job.custom_serializers << MoneySerializer
```

You won't be able to edit and reload, but you could not do it before anyway.

**Comment by notapatch** [comment]
Thank you for the prompt reply, from the Zeitwerk developer no less, and thank you for your work on the loader. That code stops the warning. If remove the autoloaders directory change the warning also stops:

```ruby
# config/initializer/custom_serializer.rb

require "money_serializer"
Rails.application.config.active_job.custom_serializers << MoneySerializer
```

I guess it's loaded it so it's not going to load it again? But it might break later? Regardless, I can confirm the workaround stop the warning. My only issue, and there's no rush, is:

-  if/when/what I need to change in the code to avoid an upgrade error.

If everything is unknown, and the changes sound complicated, it's probably best to keep the warning in my program as a reminder to future me.

**Comment by fxn** [comment]
Yes, that works because when Zeitwerk visits `app/serializers`, it detects that `MoneySerializer` is already present in memory, and therefore ignores `money_serializer.rb`. This is roughly similar to having two files with the same basename in `$LOAD_PATH`, only the first one will be loaded.

The `ignore` call would be more intentionial. You are saying clearly to Zeitwerk and to readers of the code that from `app/serializers` should be ignored as far as loading is concerned, you take care of loading it.

No need to worry about the future, this approach should work in future versions. What won't work is autoloading during boot, and your new code no longer does.

**Comment by notapatch** [comment]
Right, making it intentional makes mores sense. 👍 
 

*[re-reads guide on autoloading](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html)*

> 6 Reloading
> Rails automatically reloads classes and modules if application files change.
> 
> More precisely, if the web server is running and application files have been modified, Rails unloads all autoloaded constants just before the next request is processed. That way, application classes or modules used during that request are going to be autoloaded, thus picking up their current implementation in the file system.
> 

I have a question on `Rails.autoloaders.main.ignore` and I can't find any public documentation ([an outstanding documentation change didn't help me](https://github.com/rails/rails/pull/36102)).

To confirm, you say _"what won't work is autoloading during boot"_ so you mean that this change will only effect the boot but during later requests/response cycles the serialization classes will still be unloaded and, if needed, autoloaded again? Because the name `Rails.autoloaders.main.ignore` would make me guess it would completely remove `serializers` directory from autoloading.


### References

[Autoloading and Reloading Constants (Zeitwerk Mode)](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html#rails-autoloaders)
[Zeitwerk integration in Rails 6 (Beta 2)]( https://weblog.rubyonrails.org/2019/2/22/zeitwerk-integration-in-rails-6-beta-2/)
[How to ignore a folder in Zeitwerk for Rails 6?](https://stackoverflow.com/questions/58125879/how-to-ignore-a-folder-in-zeitwerk-for-rails-6)

**Comment by fxn** [comment]
> I have a question on Rails.autoloaders.main.ignore and I can't find any public documentation (an outstanding documentation change didn't help me).

Yes, the documentation in the guide mainly covers mostly the interface that is common to `:classic` and `:zeitwerk` modes. But if the application configures `zeitwerk` mode, then it is documented that the backend is Zeitwerk, and that you can access the individual instances using public API for further customization ([here](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html#rails-autoloaders)).

However, I've seen this same pattern a few times, so it is probably going to deserve explicit docs somewhere.

> To confirm, you say "what won't work is autoloading during boot" so you mean that this change will only effect the boot but during later requests/response cycles the serialization classes will still be unloaded and, if needed, autoloaded again? Because the name Rails.autoloaders.main.ignore would make me guess it would completely remove serializers directory from autoloading.

* When the application boots, it is a a bad idea to autoload code. That is, referencing a constant without a `require` call so that Rails goes and loads its file for you.
* In the original version of the code, the reason is the class object `MoneySerializer` evaluates to in the initializer is stored in a certain collection. It is important to understand the constant is NOT stored, it is the value it evaluates to when that initializer runs.
* When you reload, `MoneySerializer` evaluates to a different class object. But the class object cached during boot is not replaced by the new one, and the initializers do not run on reload. So the collection is storing a stale object, and therefore the changes are not reflected.

With the recommended approach you do not autoload during boot, which is the best practice. But changes won't be reflected either, because the fact that the collection is storing the class object to which `MoneySerializer` evaluated to during boot has not changed.

Bottom line, serializers are not reloadable in either case. But the original code had an issue you were not aware of, and the the warning has raised your attention to it, which is the point of the warning.

**Comment by notapatch** [comment]
OK, that took a while to digest :-) I had not even thought about autoloading before this conversation as "everything just worked". I think I'm good:
- the serializer is run at boot time when autoloading isn't reliable.
-  Serializer code runs initializer code and saves the result - trying to reload/re-run this later doesn't work.

So we are left with the problem that when someone is using the docs they are going to get a twelve line warning the solution of which has not been documented within Rails.

https://edgeguides.rubyonrails.org/active_job_basics.html#serializers

### Currently
```ruby
Rails.application.config.active_job.custom_serializers << MoneySerializer
```

### Suggested?

```ruby
# config/initializer/custom_serializer.rb

Rails.autoloaders.main.ignore("#{Rails.root}/app/serializers")

require "money_serializer"
Rails.application.config.active_job.custom_serializers << MoneySerializer
```
 
I suspect that this will lead to more questions.

**Comment by fxn** [comment]
You are good! Those docs should be revised.

**Comment by fxn** [comment]
Reopening, this probably needs a doc patch. To be evaluated.

**Comment by fxn** [comment]
This issue is not forgotten, I'll get to it.

**Comment by tonytonyjan** [comment]
Seems like suggested solution is:

```ruby
# config/initializer/custom_serializer.rb

Rails.autoloaders.main.ignore("#{Rails.root}/app/serializers")
require "money_serializer"
Rails.application.config.active_job.custom_serializers << MoneySerializer
```

I wonder if it is a workaround or the right way to do it.

**Comment by der-flo** [comment]
As a solution I could imagine that `Rails.application.config.active_job.custom_serializers` accepts classes as symbols and strings besides constants. But maybe that's too hacky…

**Comment by fxn** [comment]
This will be addressed as part of the removal of the `classic` autoloader in Rails 7.

**Comment by fxn** [comment]
This has been addressed via documentation in #41849.

Since `classic` is removed from Rails 7, I'll mention this use case in migration tips.

**Comment by bmulholland** [comment]
As noted [here](https://github.com/rails/rails/pull/41849#issuecomment-939589211), I don't think this is resolved for namespaced serializers.

**Comment by kuboon** [comment]
This worked fine. also in namespace.

```
Rails.application.reloader.to_prepare do
  Rails.application.config.active_job.custom_serializers << Money::MoneySerializer
end
```

**Comment by leoshaw** [comment]
> This worked fine. also in namespace.
> 
> ```
> Rails.application.reloader.to_prepare do
>   Rails.application.config.active_job.custom_serializers << Money::MoneySerializer
> end
> ```

We're doing this, but getting "undefined method << for nil:NilClass" when reloading because `Rails.application.config.active_job.custom_serializers` is nil when the block is evaluated.  Have looked into using `config.autoload_once_paths` to load the serializer, but it's a bit more complicated because the serializer references the class it's serializing and we want that to be autoloaded.  Currently working around it with `Rails.application.config.active_job.custom_serializers ||= []` but that doesn't feel like a good solution!  Any suggestions?
